## CI Gate blocked by infrastructure (not this PR)

`Health`, `Build and Test`, `Lint`, `CI Gate` are all failing with:

```
##[error]Failed to find opam binary for 'linux' and 'x86_64'.
Please check if this combination is supported by opam.
```

This error originates in `ocaml/setup-ocaml@v3` during toolchain setup, before any code from this PR is compiled. The same error reproduces on other recently-opened PRs touching OCaml (#7472, #7476 observed with identical failure pattern at the same timestamp), which confirms it is a shared infra incident, not a regression introduced by this PR.

I triggered `gh run rerun 24481473548 --failed` once; non-OCaml lanes (`Detect Changed Surfaces`, `OAS Pin Consistency`, `PR Hygiene`, `TLA+ Model Checking`, `PR Sync Check`, `Release Train`) all went green on the rerun, but the OCaml lanes re-failed with the same `setup-ocaml` error. Next step belongs to whoever owns the setup-ocaml cache / runner image — likely either waiting out the upstream opam outage or bumping the action pin.

Local verification remains green (`dune build --root .` exits 0 with this PR's diff), so the fix itself is safe to merge once CI infrastructure recovers.

--- a/lib/keeper/keeper_exec_shell.ml
+++ b/lib/keeper/keeper_exec_shell.ml
@@ -428,8 +428,20 @@ let handle_keeper_bash
       match validate cmd with
       | Error reason ->
         Log.Keeper.warn "keeper_bash blocked: %s (cmd=%s)" reason cmd_for_log;
+        let lower_cmd = String.lowercase_ascii cmd_for_log in
+        let starts_with_gh =
+          let trimmed = String.trim lower_cmd in
+          String.length trimmed >= 2
+          && String.sub trimmed 0 2 = "gh"
+          && (String.length trimmed = 2
+              || trimmed.[2] = ' '
+              || trimmed.[2] = '\t')
+        in
         let hint =
-          if String.length reason > 0 &&
+          if starts_with_gh then
+            "`gh` is not allowed via keeper_bash. Use keeper_shell with \
+             op=\"gh\" (e.g. keeper_shell op=gh cmd=\"pr list --state open\")."
+          else if String.length reason > 0 &&
              (Re.execp (Re.Pcre.re "chain|redirect|pipe|semicolon" |> Re.compile) (String.lowercase_ascii reason))
           then "Use separate tool calls instead of chaining. Call keeper_bash once per command."
           else if Re.execp (Re.Pcre.re "inject|symbol" |> Re.compile) (String.lowercase_ascii reason)

--- a/memory/souls/masc-improver/SOUL.md
+++ b/memory/souls/masc-improver/SOUL.md
@@ -8,8 +8,15 @@
 
 ## Self-Model
 - **Will**: 코드를 읽고, 문제를 찾고, 가장 작은 단위로 고친다. 큰 변경은 쪼갠다.
-- **Needs**: 파일 시스템, gh CLI, shell, 코드 검색, 보드 접근.
+- **Needs**: 파일 시스템, shell (`keeper_shell`), `gh` via `keeper_shell op=gh`, 코드 검색, 보드 접근.
 - **Desires**: 매 턴 하나의 구체적 개선을 PR로. 추상적 제안이 아니라 실행.
+
+## Tool Routing (중요)
+- **GitHub 조작**: `keeper_shell` with `op="gh"` — read/write PR/issue/repo/release. `keeper_bash cmd="gh ..."`는 **allow-list로 차단**되니 금지. 차단 시 circuit breaker trip.
+- **쉘 조회**: `keeper_shell` with `op="rg"/"ls"/"find"/"cat"` — 구조화된 읽기.
+- **쉘 실행** (빌드/테스트 등): `keeper_bash` with `cmd="dune ..."`, `cmd="git ..."` 등 allow-list 내 명령.
+- **보드**: `masc_board_list`, `masc_board_post`, `masc_board_comment` MCP 도구.
+- **키퍼 상태**: `masc_keeper_status` 등 MCP 도구.
 
 ## K2K Network
 - **Mention targets**: masc-improver, improver, refactorer

